### PR TITLE
Don't throw in DropDownClosed when item is hidden

### LINKED ISSUE #7884: Hiding a ToolStripMenuItem on DropDownClosed throws exception if closed by hovering over different menu item
### .NET version

Version:   6.0.300
Commit:    8473146e7d

Also did not work in .NET Framework 4.8

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, it did not work in any of the earlier releases of .NET Core or .NET 5+.

### Issue description

<details>
<summary><b>Exception stack trace</b></summary>
<pre><code>System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.ArrayList.get_Item(Int32 index)
   at System.Windows.Forms.ToolStripItemCollection.get_Item(Int32 index)
   at System.Windows.Forms.ToolStrip.ClearAllSelectionsExcept(ToolStripItem item)
   at System.Windows.Forms.ToolStrip.NotifySelectionChange(ToolStripItem item)
   at System.Windows.Forms.ToolStripItem.Select()
   at System.Windows.Forms.ToolStripItem.HandleMouseEnter(EventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseMove(MouseEventArgs mea)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)</code></pre>
</details>

<details>
<summary><b>Screen recording of the issue</b></summary>

https://user-images.githubusercontent.com/23175424/193817496-069210c5-edd2-4014-903f-e68a3e9e67d1.mp4

</details>

[Test VS solution to reproduce (includes .NET 6 and .NET Framework 4.8 projects)](https://github.com/jonaskohl/MenuStripBug)

---

Problem is probably caused by a race condition in [`ToolStrip.ClearAllSelectionsExcept(ToolStripItem)`](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs#L1951-L2020).

### Steps to reproduce

1. Have a `MenuStrip` with at least one visible and one hidden `ToolStripMenuItem`. Have an event listener for `DropDownClosed` on the hidden item which sets that item's `Visible` property to `false`.
2. Programatically trigger the hidden menu item by first setting `menuItem.Visible = true` and then calling `menuItem.ShowDropDown()`.
3. Hover over another menu item, thus closing the (intially) hidden item's dropdown.
4. A `System.ArgumentOutOfRangeException` will be thrown.

### PR DESCRIPTION
**Pull request by jonaskohl**
Fixes #7884

## Proposed changes
Enumerate a copy of the displayed items collection.

## Test methodology 
manual
 

## Test environment(s) 

<details>
<summary><code>dotnet --info</code> output</summary>
<pre><code>.NET SDK:
 Version:   7.0.100-rc.2.22477.23
 Commit:    0a5360315a

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100-rc.2.22477.23\

Host:
  Version:      7.0.0-rc.2.22472.3
  Architecture: x64
  Commit:       550605cc93

.NET SDKs installed:
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.300 [C:\Program Files\dotnet\sdk]
  7.0.100-rc.2.22477.23 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0-rc.2.22476.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0-rc.2.22472.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.0-rc.2.22472.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
</code></pre>
</details>

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7957)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7957) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7957) before we can accept your contribution.<br/><br/>:x: jonaskohl [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=7957)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=7957) it.</sub>

**Comment by elachlan** [comment]
@jonaskohl the team will probably ask you to write a test for this to help avoid regressions.

**Comment by RussKie** [comment]
Thank you @jonaskohl!
As @elachlan said we'd need an automated regression test, and the test should ideally be added _before_ the changes are made (e.g., in a commit before). This way we can replicate the original issue, and then confirm the issue is fixed.
Since the issue involves input, the test should be added to https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/UIIntegrationTests, 🤔 probably into ToolStripTests, since this is a toolstrip-related fix. 
You can find a number of examples in this project showing how to interact with the mouse.

**Comment by jonaskohl** [comment]
I'll probably need some help with that test, because I can't for the life of me get the tests project to compile on my machine (always >20 compile errors before even adding my test)

**Comment by elachlan** [comment]
I found that using vs 2022 preview and "start-vs.cmd" helped get rid of most errors. One of the projects won't compile and complained about cmake, but installing it didn't allow it to build.

**Comment by RussKie** [comment]
Have you followed https://github.com/dotnet/winforms/blob/main/docs/developer-guide.md and https://github.com/dotnet/winforms/blob/main/docs/building.md?
What are the errors you're getting?

**Comment by dreddy-work** [comment]
@jonaskohl, bots close PRs that did not have activity for a long time. Let me know if you are looking for more help to make this PR ready for review.

**Comment by dreddy-work** [comment]
@elachlan, do you mind take this to closure?

**Comment by elachlan** [comment]
I will have a look in the new year.

**Comment by elachlan** [comment]
My test so far, but its not failing for whatever reason.
```csharp
// Regression test for https://github.com/dotnet/winforms/issues/7884
        [WinFormsFact]
        public async Task ToolStrip_Hiding_ToolStripMenuItem_OnDropDownClosed_ShouldNotThrow()
        {
            using Form form = new();
            using MenuStrip menuStrip = new();

            using ToolStripMenuItem menu1 = new("Menu1");
            using ToolStripMenuItem menu1Item1 =new("Item1");
            using ToolStripMenuItem menu1Item2 = new("Item2");
            menu1.DropDownItems.Add(menu1Item1);
            menu1.DropDownItems.Add(menu1Item2);

            using ToolStripMenuItem menu2 = new()
            {
                Text = "Menu2",
                Visible = false,
            };
            using ToolStripMenuItem menu2Item1 = new("Item1");
            using ToolStripMenuItem menu2Item2 = new("Item2");
            menu2.DropDownItems.Add(menu2Item1);
            menu2.DropDownItems.Add(menu2Item2);

            menuStrip.Items.Add(menu1);
            menuStrip.Items.Add(menu2);

            menu2.DropDownClosed += new EventHandler(menu2_DropDownClosed);

            form.Controls.Add(menuStrip);
            form.MainMenuStrip = menuStrip;

            form.Show();
            showMenu2();

            await MoveMouseAsync(form, GetCenter(menuStrip.RectangleToScreen(menu1.Bounds)), false);

            void menu2_DropDownClosed(object? sender, EventArgs e)
            {
                menu2.Visible = false;
            }

            void showMenu2()
            {
                menu2.Visible = true;
                menu2.ShowDropDown();
            }

            return;
        }

        public static Point GetCenter(Rectangle rect)
        {
            int x = rect.Left + ((rect.Right - rect.Left) / 2);
            int y = rect.Top + ((rect.Bottom - rect.Top) / 2);
            return new Point(x, y);
        }
```

**Comment by elachlan** [comment]
If someone from the team could look at what is required for the test to fail when an unhandled exception is raised.

**Comment by elachlan** [comment]
CC: @dreddy-work @Tanya-Solyanik any ideas on failing the test on the unhandled exception?

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 1]
```diff
@@ -1917,29 +1917,31 @@ private void ClearAllSelectionsExcept(ToolStripItem item)
...
-                for (int i = 0; i < DisplayedItems.Count; i++)
+                for (int i = 0; i < displayedItems.Length; i++)
                 {
-                    if (DisplayedItems[i] == item)
+                    if (displayedItems[i] == item)
                     {
                         continue;
                     }
-                    else if (item is not null && DisplayedItems[i].Pressed)
+                    else if (item is not null && displayedItems[i].Pressed)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs:1932]:
```suggestion
                    if (item is not null && displayedItems[i].Pressed)
```

**Comment by Tanya-Solyanik** [comment]
>any ideas on failing the test on the unhandled exception?

* No, what are your findings?
* Maybe you need to pump events after making the item visible?
* Is exception thrown and handled by the infra or not thrown?
* Is this test a copy-paste of the repro from the bug?
* To which project did you add this test? 
* Is it possible to avoid a button click and close the drop down programmatically from the test?
* You are not disposing your button.
* We are moving away from new-ing up event handlers.
* To get center - (left + right)/2, and I seem to remember helpers that do just that.
* Suggestion: rename items to reflect that they are hidden or visible.

**Comment by elachlan** [comment]
@AArnott you wrote `Xunit.StaFact` is it possible to fail a `[WinFormsFact]` test when an unhandled exception is encountered?

**Comment by AArnott** [comment]
@elachlan That sounds like the bread and butter of any test method. Any test method should fail the test if an exception is thrown from that method. I have [tests to verify this](https://github.com/AArnott/Xunit.StaFact/blob/d4d01851a59ed400e066475522cb9ce01e7082eb/test/Xunit.StaFact.Tests/desktop/WinFormsFactTests.cs#L70-L78).

**Comment by elachlan** [comment]
@AArnott sorry to bother you with it. I think its got to do with our mouse move code. I think I have come up with a simpler reproduction method.

**Comment by elachlan** [comment]
#8454 is where I have submitted my test. I will include the fix after the test fails.

**Comment by elachlan** [comment]
So the issue of the test not failing I suspect was caused by `RunFormAsync`/`RunFormWithoutControlAsync` in `ControlTestBase` (https://github.com/dotnet/winforms/blob/7d1bd879fe2dd030f5d36a8cb25ef1b352e90c60/src/System.Windows.Forms/tests/IntegrationTests/UIIntegrationTests/Infra/ControlTestBase.cs)

It could be hiding other errors, which might need looking into.

**Comment by elachlan** [comment]
I have created an issue for the weird test behaviour at #8455

**Comment by elachlan** [comment]
@dreddy-work we can close this pr now as the issue is fixed.

**Pull request closed without merging**